mlir/lib/Dialect/Transform/IR/TransformOps.cpp

```diff
@@ -796,6 +800,294 @@ DiagnosedSilenceableFailure transform::ApplyRegisteredPassOp::applyToOne(
+
+      // Set this result as the next result of the scope op.
+      liveOuts.push_back(result);
+      liveOutIndex[result] = nextScopeResultIndex++;
+      resultTypes.push_back(result.getType());
+    }
+  }
+
+  if (where == RelativeLocation::Before)
+    rewriter.setInsertionPoint(wherePayload);
+  else
+    rewriter.setInsertionPointAfter(wherePayload);
+
+#define WRAP_IN_EXECUTE_REGION
+#ifdef WRAP_IN_EXECUTE_REGION
+  TypeRange noTypes;
+  ValueRange noValues;
+  auto executeRegion = rewriter.create<scf::ExecuteRegionOp>(
+      wherePayload->getLoc(), noTypes, noValues);
+  Block &executeRegionBody = executeRegion.getRegion().emplaceBlock();
+  rewriter.setInsertionPointToStart(&executeRegionBody);
+#endif
+
+  auto scope = rewriter.create<scf::ScopeOp>(wherePayload->getLoc(),
+                                             resultTypes, liveIns);
+  Region *scopeBody = &scope.getBody();
+  // TODO: Move into builder.
+  Block *scopeBlock = rewriter.createBlock(scopeBody, scopeBody->end());
+
+  // TODO: Move into builder.
+  for (Value arg : liveIns)
+    scopeBlock->addArgument(arg.getType(), arg.getLoc());
+
+  IRMapping mapper;
+  for (Value liveIn : liveIns)
+    mapper.map(liveIn, scopeBlock->getArgument(liveInIndex[liveIn]));
+
+  rewriter.setInsertionPointToEnd(scopeBlock);
+  for (auto payload : opsPayload)
+    rewriter.clone(*payload, mapper);
+
+  SmallVector<Value> scopeResults;
+  for (Value liveOut : liveOuts)
+    scopeResults.push_back(mapper.lookup(liveOut));
+  rewriter.create<scf::YieldOp>(scope.getLoc(), scopeResults);
+
+  Region &body = getBody();
+
+  DiagnosedSilenceableFailure result = DiagnosedSilenceableFailure::success();
+
+  // Create a region scope for the transform::as_scope region and run the
+  // transformations it contains. This is done in a syntactic C++ scope since
+  // we'll be deleting the scf::scope, and its mapping will become invalid if
+  // left until compactOpHandles() is called.
+  {
+    auto regionScope = state.make_region_scope(body);
+    if (failed(state.mapBlockArguments(body.front().getArgument(0), {scope})))
+      return DiagnosedSilenceableFailure::definiteFailure();
+
+    for (Operation &transform : body.front().without_terminator()) {
+      result = state.applyTransform(cast<TransformOpInterface>(transform));
+      if (result.isSilenceableFailure()) {
+        LLVM_DEBUG(DBGS() << "transformation on scope failed: "
+                          << result.getMessage() << "\n");
+        break;
+      }
+
+      if (::mlir::failed(result.silence()))
+        return DiagnosedSilenceableFailure::definiteFailure();
+    }
+
+#ifdef WRAP_IN_EXECUTE_REGION
+    // Since the transformations applied may have replaced the scope, get the
+    // current scope from the wrapping scf.execute_region.
+    scope = dyn_cast<scf::ScopeOp>(*executeRegionBody.getOperations().begin());
+    scope->moveAfter(executeRegion);
+    rewriter.eraseOp(executeRegion);
+#else
```

Comment:
The (simpler) code for retrieving the (potentially replaced) scf.scope after applying the nested transforms in the #else clause fails for transform.alternatives whose getEffects() consumes the handle it is given (not sure why, as it seems to expect the clones to exist after applying the alternative, @ftynse could you explain?). I resorted to wrapping the scf.scope in a scf.execute_region, but there must be a more elegant way to do this.

Reply:
My hypothesis is the following. The `auto newPayloadOps = state.getPayloadOps(body.front().getArgument(0));` line below gets the payload ops associated with the block argument handle by `mapBlockArguments` above, that is, `scope`. The nested transforms ops may or may not update that association. Specifically, if the scope is rewritten/deleted, the `Operation *` contained in this association will be dangling (but may happen to point to another operation, including the scope that was created as a replacement, by accident of the allocator reusing the pointer address). Hopefully, some of the sanity checks in place will complain that access to a consumed handle is requested (if they don't, let me know and we'll try to ensure they do). The workaround above gets the scope operation directly from the surrounding `execute_region`, which is never rewritten, rather than from the handle association, and is therefore immune to such changes.

The solution to this is to `transform.yield` the (updated) scope handle from the `transform.as_scope` and use the payload associated with the terminator operand to get the most recent scope. Some more thinking is needed to make this interop with `transform.alternative` as that op doesn't return the scope, not sure whether it should return the updated scope or mark the handle as consuming only nested payloads but not the handle itself. I don't think we have a mechanism for the latter, but the need for it was floated around.

Reply:
keeping the wrapping op solution for now, as this seems to require a broader discussion.